vertx-mongo-client/src/main/java/io/vertx/ext/mongo/impl/MongoIterableStream.java

```diff
@@ -0,0 +1,186 @@
+package io.vertx.ext.mongo.impl;
+
+import com.mongodb.async.AsyncBatchCursor;
+import com.mongodb.async.SingleResultCallback;
+import com.mongodb.async.client.MongoIterable;
+import io.vertx.core.Context;
+import io.vertx.core.Handler;
+import io.vertx.core.json.JsonObject;
+import io.vertx.core.streams.ReadStream;
+
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * @author Thomas Segismont
+ */
+class MongoIterableStream implements ReadStream<JsonObject> {
+
+  private static final int BATCH_SIZE = 10;
+
+  private final Context context;
+  private final MongoIterable<JsonObject> mongoIterable;
+
+  private AsyncBatchCursor<JsonObject> batchCursor;
+  private Deque<JsonObject> queue;
+  private Handler<JsonObject> dataHandler;
+  private Handler<Throwable> exceptionHandler;
+  private Handler<Void> endHandler;
+  private boolean paused;
+  private boolean readInProgress;
+  private boolean closed;
+
+  MongoIterableStream(Context context, MongoIterable<JsonObject> mongoIterable) {
+    this.context = context;
+    this.mongoIterable = mongoIterable;
+  }
+
+  @Override
+  public synchronized MongoIterableStream exceptionHandler(Handler<Throwable> handler) {
+    checkClosed();
+    this.exceptionHandler = handler;
+    return this;
+  }
+
+  private void checkClosed() {
+    if (closed) {
```

Comment:
For safety, should be closed mark as volatile or the method also synchronized.
I understand that all access are made when holding the monitor lock, it's just in case....

Reply:
I prefer not to make it volatile. As you noticed, it's guarded by the stream instance, and we'll ruin the benefits of biased locking otherwise.

Reply:
Worth adding a comment to that effect